akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala

```diff
@@ -269,6 +270,59 @@ object GraphStages {
   def terminationWatcher[T]: GraphStageWithMaterializedValue[FlowShape[T, T], Future[Done]] =
     TerminationWatcher.asInstanceOf[GraphStageWithMaterializedValue[FlowShape[T, T], Future[Done]]]
 
+  private class FlowMonitorImpl[T] extends AtomicReference[Any](Initialized) with FlowMonitor[T] {
+    override def state = get match {
+      case s: StreamState[_] ⇒ s.asInstanceOf[StreamState[T]]
+      case msg               ⇒ Received(msg.asInstanceOf[T])
+    }
+  }
+
+  private class MonitorFlow[T] extends GraphStageWithMaterializedValue[FlowShape[T, T], FlowMonitor[T]] {
+    val in = Inlet[T]("FlowMonitor.in")
+    val out = Outlet[T]("FlowMonitor.out")
+    val shape = FlowShape.of(in, out)
+
+    override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, FlowMonitor[T]) = {
+      val monitor: FlowMonitorImpl[T] = new FlowMonitorImpl[T]
+
+      val logic: GraphStageLogic = new GraphStageLogic(shape) {
+        setHandler(in, new InHandler {
+          override def onPush(): Unit = {
+            val msg = grab(in)
+            push(out, msg)
+            monitor.set(if (msg.isInstanceOf[StreamState[_]]) Received(msg) else msg)
+          }
+          override def onUpstreamFinish(): Unit = {
+            super.onUpstreamFinish()
+            monitor.set(Finished)
+            completeStage()
```

Comment:
yes, they do. I had here a wrong comment which I deleted immediately after. Have you been fixing my ghost comment :)